src/test/scala/com/mongodb/spark/sql/MapFunctionsSpec.scala

```diff
@@ -0,0 +1,144 @@
+    row.toSeq should equal(Array("John", null))
+    row.schema should equal(schema)
+  }
+
+  it should "prune the schema when limited by passed required columns" in {
+    val schema: StructType = schemaFor[Person]
+    val document: BsonDocument = BsonDocument.parse("{name: 'John', age: 18}")
+
+    val row: Row = documentToRow(document, schema, Array("age"))
+    row.toSeq should equal(Array(18))
+    row.schema should equal(DataTypes.createStructType(Array(schema.fields(1))))
+  }
+
+  it should "ignore any extra data in the document that is not included in the schema" in {
+    val schema: StructType = schemaFor[Person]
+    val document: BsonDocument = BsonDocument.parse("{name: 'John', age: 18, height: 192}")
+
+    val row: Row = documentToRow(document, schema)
+    row.toSeq should equal(Array("John", 18))
+    row.schema should equal(schema)
+  }
+
+  it should "handle nested schemas" in {
+    val schema: StructType = schemaFor[Family]
+    val document: BsonDocument = BsonDocument.parse(
+      """{familyName: "Smith", members:[
+        |{name: "James", age: 48},
+        |{name: "Jane", age: 42},
+        |{name: 'Jeremy', age: 18},
+        |{name: 'John', age: 18}]}""".stripMargin
+    )
+
+    val row: Row = documentToRow(document, schema)
+    row.getAs[String]("familyName") should equal("Smith")
+    row.getSeq[Row](1).map(_.toSeq) should equal(Array(Array("James", 48), Array("Jane", 42), Array("Jeremy", 18), Array("John", 18)))
+    row.schema should equal(schema)
+  }
+
+  "rowToDocument" should "convert a Row into a Document" in {
+    val schema: StructType = schemaFor[Person]
+    val original: BsonDocument = BsonDocument.parse("{name: 'John', age: 18}")
+
+    val row: Row = documentToRow(original, schema)
+    val converted: BsonDocument = rowToDocument(row)
+
+    converted should equal(original)
+  }
+
+  it should "handle nested schemas" in {
+    val schema: StructType = schemaFor[Family]
+    val original: BsonDocument = BsonDocument.parse(
+      """{familyName: "Smith", members:[
+        |{name: "James", age: 48},
+        |{name: "Jane", age: 42},
+        |{name: 'Jeremy', age: 18},
+        |{name: 'John', age: 18}]}""".stripMargin
+    )
+
+    val row: Row = documentToRow(original, schema)
+    val converted: BsonDocument = rowToDocument(row)
+
+    converted should equal(original)
+  }
+
+  it should "handle nested schemas within nested arrays" in {
+    val schema: StructType = schemaFor[NestedFamily]
+    val original: BsonDocument = BsonDocument.parse(
+      """{familyName: "Smith", members:[
+        |[{name: "James", age: 48}, {name: "Jane", age: 42}],
+        |[{name: 'Jeremy', age: 18}, {name: 'John', age: 18}]]}""".stripMargin
+    )
+
+    val row: Row = documentToRow(original, schema)
+    val converted: BsonDocument = rowToDocument(row)
+
+    converted should equal(original)
+  }
+
+  it should "handle mixed numerics based on the schema" in {
```

Comment:
By using BsonDocument.parse, this test is relying on unspecified behavior for MongoDB extended JSON:  that 1 parses to a BsonInt32 and 1.0 parses to a BsonDouble.  It would be clearer to not use parse, e.g.

  new BsonDocument("num", new BsonInt32(1))
  new BsonDocument("num", new BsonDouble(1))